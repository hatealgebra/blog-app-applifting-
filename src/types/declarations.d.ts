export interface paths {
  '/login': {
    post: {
      responses: {
        /** Access token detail */
        201: {
          content: {
            'application/json': components['schemas']['AccessToken'];
          };
        };
        /** Invalid login credentials */
        400: {
          content: {
            'application/json': unknown;
          };
        };
        401: components['responses']['ApiKeyInvalidError'];
      };
      requestBody: {
        content: {
          'application/json': {
            username?: string;
            /** Format: password */
            password: string;
          };
        };
      };
    };
  };
  '/articles': {
    get: operations['listArticles'];
    post: operations['createArticle'];
  };
  '/articles/{articleId}': {
    get: operations['getArticle'];
    delete: operations['deleteArticle'];
    patch: operations['updateArticle'];
  };
  '/comments': {
    post: {
      responses: {
        /** Detail of the created comment */
        201: {
          content: {
            'application/json': components['schemas']['Comment'];
          };
        };
        401: components['responses']['ApiKeyInvalidError'];
      };
      requestBody: {
        content: {
          'application/json': components['schemas']['Comment'];
        };
      };
    };
  };
  '/comments/{commentId}/vote/up': {
    post: {
      parameters: {
        path: {
          /** Id of an comment */
          commentId: components['parameters']['commentId'];
        };
      };
      responses: {
        /** Updated comment detail */
        201: {
          content: {
            'application/json': components['schemas']['Comment'];
          };
        };
        401: components['responses']['ApiKeyInvalidError'];
      };
    };
  };
  '/comments/{commentId}/vote/down': {
    post: {
      parameters: {
        path: {
          /** Id of an comment */
          commentId: components['parameters']['commentId'];
        };
      };
      responses: {
        /** Updated comment detail */
        201: {
          content: {
            'application/json': components['schemas']['Comment'];
          };
        };
        401: components['responses']['ApiKeyInvalidError'];
      };
    };
  };
  '/images': {
    post: {
      responses: {
        /** Image uploaded successfully */
        201: {
          content: {
            'application/json': components['schemas']['ImageInfo'][];
          };
        };
        401: components['responses']['ApiKeyInvalidError'];
      };
      requestBody: {
        content: {
          'multipart/form-data': {
            image?: string[];
          };
        };
      };
    };
  };
  '/images/{imageId}': {
    get: {
      parameters: {
        path: {
          /** Id of the image to retrieve */
          imageId: components['parameters']['imageId'];
        };
      };
      responses: {
        /** Image file */
        200: unknown;
        401: components['responses']['ApiKeyInvalidError'];
      };
    };
    delete: {
      parameters: {
        path: {
          /** Id of the image to retrieve */
          imageId: components['parameters']['imageId'];
        };
      };
      responses: {
        /** Image no longer exists */
        204: never;
        401: components['responses']['ApiKeyInvalidError'];
      };
    };
  };
  '/tenants': {
    post: {
      responses: {
        /** Detail of the created tenant */
        201: {
          content: {
            'application/json': components['schemas']['Tenant'];
          };
        };
      };
      requestBody: {
        content: {
          'application/json': components['schemas']['Tenant'];
        };
      };
    };
  };
  '/tenants/{tenantId}': {
    get: {
      parameters: {
        path: {
          /** Id of the tenant to retrieve */
          tenantId: components['parameters']['tenantId'];
        };
      };
      responses: {
        /** Tenant detail */
        200: {
          content: {
            'application/json': components['schemas']['Tenant'];
          };
        };
      };
    };
  };
}

export interface components {
  schemas: {
    Pagination: {
      /** @description Number of items skipped during pagination */
      offset?: number;
      /** @description Number of items returned */
      limit?: number;
      /** @description Total number of items */
      total?: number;
    };
    ArticleList: {
      pagination?: components['schemas']['Pagination'];
      items?: components['schemas']['Article'][];
    };
    Article: {
      /** Format: uuid */
      articleId?: string;
      /** @example Lorem Ipsum */
      title?: string;
      /** @example Lorem Ipsum is simply dummy text of the printing and typesetting industry. */
      perex?: string;
      /** Format: uuid */
      imageId?: string;
      /** Format: date-time */
      createdAt?: string;
      /** Format: date-time */
      lastUpdatedAt?: string;
    };
    ArticleDetail: components['schemas']['Article'] & {
      /**
       * Format: markdown
       * @example # Lorem Ipsum
       * **Lorem Ipsum** is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
       */
      content?: string;
      comments?: components['schemas']['Comment'][];
    };
    Comment: {
      /** Format: uuid */
      commentId?: string;
      /** Format: uuid */
      articleId: string;
      /** @example Jan Novak */
      author: string;
      /** @example This article is on point and very inspirational. */
      content: string;
      /** Format: date-time */
      postedAt?: string;
      /** @example 42 */
      score?: number;
    };
    ImageInfo: {
      /** Format: uuid */
      imageId: string;
      /** @example screenshot.png */
      name?: string;
    };
    Tenant: {
      /** Format: uuid */
      tenantId?: string;
      /**
       * Format: uuid
       * @description API key to be used in the `api-key` header
       */
      apiKey?: string;
      /**
       * @description Human readable name of the tenant
       * @example Jan Novak
       */
      name?: string;
      /**
       * Format: password
       * @description Authentication password of the tenant
       * @example correct horse battery staple
       */
      password?: string;
      /** Format: date-time */
      createdAt?: string;
      /** Format: date-time */
      lastUsedAt?: string;
    };
    AccessToken: {
      /**
       * Format: uuid
       * @description Use this when requesting authenticated API endpoints
       */
      access_token?: string;
      /**
       * @description Number of seconds before the access_token expires
       * @example 3600
       */
      expires_in?: number;
      /**
       * @description Token type. Will always be bearer
       * @example bearer
       */
      token_type?: string;
    };
  };
  responses: {
    /** API key is missing or invalid */
    ApiKeyInvalidError: {
      content: {
        'application/json': unknown;
      };
    };
    /** Access token is missing or invalid */
    UnauthorizedError: {
      content: {
        'application/json': unknown;
      };
    };
  };
  parameters: {
    /** @description Id of the tenant to retrieve */
    tenantId: string;
    /** @description Id of the image to retrieve */
    imageId: string;
    /** @description Id of an article */
    articleId: string;
    /** @description Id of an comment */
    commentId: string;
    /** @description Number of items to skip during pagination */
    offset: number;
    /** @description Number of items to return per page. All items are returned if `limit` is omitted. */
    limit: number;
  };
}

export interface operations {
  listArticles: {
    parameters: {
      query: {
        /** Number of items to skip during pagination */
        offset?: components['parameters']['offset'];
        /** Number of items to return per page. All items are returned if `limit` is omitted. */
        limit?: components['parameters']['limit'];
      };
    };
    responses: {
      /** Article list */
      200: {
        content: {
          'application/json': components['schemas']['Article'][];
        };
      };
      401: components['responses']['ApiKeyInvalidError'];
    };
  };
  createArticle: {
    responses: {
      /** Detail of the created article */
      200: {
        content: {
          'application/json': components['schemas']['ArticleDetail'];
        };
      };
      401: components['responses']['ApiKeyInvalidError'];
      403: components['responses']['UnauthorizedError'];
    };
    requestBody: {
      content: {
        'application/json': components['schemas']['ArticleDetail'];
      };
    };
  };
  getArticle: {
    parameters: {
      path: {
        /** Id of an article */
        articleId: components['parameters']['articleId'];
      };
    };
    responses: {
      /** Article detail */
      200: {
        content: {
          'application/json': components['schemas']['ArticleDetail'];
        };
      };
      401: components['responses']['ApiKeyInvalidError'];
    };
  };
  deleteArticle: {
    parameters: {
      path: {
        /** Id of an article */
        articleId: components['parameters']['articleId'];
      };
    };
    responses: {
      /** Article no longer exists */
      204: never;
      401: components['responses']['ApiKeyInvalidError'];
      403: components['responses']['UnauthorizedError'];
    };
  };
  updateArticle: {
    parameters: {
      path: {
        /** Id of an article */
        articleId: components['parameters']['articleId'];
      };
    };
    responses: {
      /** Updated article detail */
      200: {
        content: {
          'application/json': components['schemas']['ArticleDetail'];
        };
      };
      401: components['responses']['ApiKeyInvalidError'];
      403: components['responses']['UnauthorizedError'];
    };
    requestBody: {
      content: {
        'application/json': components['schemas']['ArticleDetail'];
      };
    };
  };
}

export interface external {}
