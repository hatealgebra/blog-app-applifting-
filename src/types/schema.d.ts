export interface IArticle {
  articleId: string;
  title: string;
  perex: string;
  imageId: string;
  createdAt: string;
  lastUpdateAt: string;
}

export interface IArticleDetail {}
