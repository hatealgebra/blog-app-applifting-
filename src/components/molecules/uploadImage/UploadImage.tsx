import React from 'react';
import Button from '../../atoms/button/Button';
import { StyledMinimalButton, StyledUploadImage } from './uploadImage.styled';

const UploadImage = ({ image, setImage }: UploadImageProps) => {
  const [imagePreview, setImagePreview] = React.useState<string | undefined>(
    null
  );
  const hiddenFileInputRef = React.useRef<HTMLInputElement>(null);

  const handleInput = () => {
    if (hiddenFileInputRef.current !== null) {
      hiddenFileInputRef.current.click();
    }
  };

  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files !== null) {
      const fileUploaded = event.target.files[0];
      setImage(fileUploaded);
    }
  };

  const resetImage = () => setImage(null);

  React.useEffect(() => {
    if (image !== null) {
      const objectURL = URL.createObjectURL(image);
      setImagePreview(objectURL);
    } else {
      setImagePreview(undefined);
    }
  }, [image]);

  return (
    <StyledUploadImage>
      <label className="upload-image__label">Featured Image</label>
      {image !== null ? (
        <>
          <img
            className="upload-image__preview"
            src={imagePreview}
            alt={image.name}
          />
          <div className="upload-image__button-row">
            <StyledMinimalButton onClick={handleInput}>
              Upload New
            </StyledMinimalButton>
            <StyledMinimalButton onClick={resetImage}>
              Delete
            </StyledMinimalButton>
          </div>
        </>
      ) : (
        <Button onClick={handleInput} colortheme="secondary">
          Upload an Image
        </Button>
      )}
      <input
        data-testid="image-uploader"
        ref={hiddenFileInputRef}
        type="file"
        style={{ display: 'none' }}
        accept="image/*"
        onChange={(e) => handleFileChange(e)}
      />
    </StyledUploadImage>
  );
};

export interface UploadImageProps {
  image: File | null;
  setImage: React.Dispatch<React.SetStateAction<File | null>>;
}

export default UploadImage;
