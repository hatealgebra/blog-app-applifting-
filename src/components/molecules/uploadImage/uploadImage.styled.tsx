import styled from 'styled-components';
import { ButtonProps } from '@atoms/button/button.types';
import Button from '@atoms/button/Button';

export const StyledMinimalButton = styled(Button).attrs<ButtonProps>(() => ({
  variant: 'outline',
  size: 'sm',
}))<ButtonProps>`
  /* stylelint-disable */
  /* stylelint-enable */
`;

export const StyledUploadImage = styled.div`
  display: flex;
  flex-direction: column;
  row-gap: 10px;

  .upload-image {
    button,
    &__preview {
      width: 112px;
      height: 74px;
    }

    &__button-row {
      display: flex;

      button {
        border: none;
        border-radius: 0;
      }

      button:nth-of-type(1) {
        color: ${({ theme }) => theme.color.primary};
        padding-left: 0;
        border-right: 2px solid ${({ theme }) => theme.color.border};
      }

      button:nth-of-type(2) {
        color: ${({ theme }) => theme.color.danger};
      }
    }
  }
`;
