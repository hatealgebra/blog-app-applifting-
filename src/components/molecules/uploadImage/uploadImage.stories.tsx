import React from 'react';
import UploadImage from './UploadImage';

export const UploadImageExample = () => {
  const [image, setImage] = React.useState<File | null>(null);
  return <UploadImage image={image} setImage={setImage} />;
};

export const UploadImageWithFile = () => {
  const file = new File(['hello'], 'hello.png', { type: 'image/png' });
  const [image, setImage] = React.useState<File | null>(file);
  return <UploadImage image={image} setImage={setImage} />;
};
export default {
  title: 'Molecules/Upload image',
  component: UploadImage,
};
