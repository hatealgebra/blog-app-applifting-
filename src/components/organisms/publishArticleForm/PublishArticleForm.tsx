import React, { FormEvent } from 'react';
import {
  updateArticleHelper,
  validatePublishArticleForm,
} from '@helpers/publishArticle.helper';
import AdminHeading from '../../molecules/adminHeading/AdminHeading';
import MarkdownEditor from '../../atoms/markdownEditor/MarkdownEditor';
import InputWithLabel from '../../molecules/inputWithLabel/InputWithLabel';
import UploadImage from '../../molecules/uploadImage/UploadImage';
import StyledPublishArticleForm from './publishArticleForm.styled';
import ErrorText from '../../atoms/errorText/error.styled';
import { useAppDispatch, useAppSelector } from '../../../store/hooks';

import { selectAuthToken } from '../../../store/slices/auth.slices';

import { cutTextWithElipsis } from '../../../utils/generic.utils';
import { createArticleThunk } from '../../../store/thunks/admin.thunks';
import {
  EPublishArticleErrors,
  PublishArticleProps,
} from './publishArticleForm.types.d';

// FIXME: maybe implement do BIG notation?
// TODO: Testing
// FIXME: When editing the article, get the loading screen maybe?
// TODO: useEffect for for error handling
const PublishArticleForm = ({
  titleValue,
  markdownContentValue,
  imageFileValue,
}: PublishArticleProps) => {
  const [articleId] = React.useState(undefined);
  const [title, setTitle] = React.useState(titleValue || '');
  const [markdownContent, setMarkdownContent] = React.useState(
    markdownContentValue || ''
  );
  const [imageFile, setImageFile] = React.useState<File | null>(
    imageFileValue || null
  );
  const [isImageChanged, setIsImageChanged] = React.useState(false);
  const [formError, setFormError] = React.useState<EPublishArticleErrors>(
    EPublishArticleErrors.PASSED
  );

  // eslint-disable-next-line @typescript-eslint/naming-convention
  const access_token = useAppSelector(selectAuthToken);
  const dispatch = useAppDispatch();

  const handleOnSubmit = (e: FormEvent) => {
    e.preventDefault();
    const trimmedTitle = title.trim();
    const trimmedMD = markdownContent.trim();
    const perex = cutTextWithElipsis(trimmedMD, 130);

    const formValidationPassed = validatePublishArticleForm(
      trimmedTitle,
      trimmedMD,
      imageFile,
      setFormError
    );

    if (formValidationPassed) {
      const imageFormData = new FormData();
      imageFormData.append('image', imageFile!);
      if (articleId) {
        return updateArticleHelper(
          e,
          articleId,
          trimmedTitle,
          perex,
          trimmedMD,
          imageFormData!,
          access_token,
          isImageChanged
        );
      }
      return dispatch(
        createArticleThunk({
          title: trimmedTitle,
          perex,
          content: trimmedMD,
          imageFormData,
          access_token,
        })
      );
    }
    return false;
  };

  React.useEffect(() => {
    setIsImageChanged(true);
  }, [setImageFile]);

  return (
    <StyledPublishArticleForm onSubmit={(e) => handleOnSubmit(e)}>
      <AdminHeading
        heading="Create new article"
        buttonText="Publish article"
        isFormPage
      />
      <div>
        <InputWithLabel
          label="Article title"
          value={title}
          onChange={(e) => setTitle(e.target.value)}
          placeholder="My new article"
        />
        <ErrorText>
          {formError === EPublishArticleErrors.TITLE_EMPTY &&
            EPublishArticleErrors.TITLE_EMPTY}
        </ErrorText>
      </div>
      <div>
        <UploadImage image={imageFile} setImage={setImageFile} />
        <ErrorText>
          {formError === EPublishArticleErrors.IMAGE_EMPTY &&
            EPublishArticleErrors.IMAGE_EMPTY}
        </ErrorText>
      </div>
      <div>
        <MarkdownEditor value={markdownContent} onChange={setMarkdownContent} />
        <ErrorText>{formError}</ErrorText>
      </div>
    </StyledPublishArticleForm>
  );
};

export default PublishArticleForm;
