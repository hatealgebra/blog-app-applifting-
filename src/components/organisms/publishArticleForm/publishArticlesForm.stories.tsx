import { expect } from '@storybook/jest';
import React from 'react';

import { userEvent, within } from '@storybook/testing-library';
import allArticlesDetailMockJSON from '@mocks/asyncData/get/articlesDetailsResponse.mock.json';
import { mockOriginalFile } from '@molecules/uploadImage/uploadImage.stories';
import PublishArticleForm from './PublishArticleForm';
import { EPublishArticleErrors } from './publishArticleForm.types.d';

const { title, content } = allArticlesDetailMockJSON.items[0];

const Template = (args) => <PublishArticleForm {...args} />;

export const CreateNewArticleFormExample = Template.bind({});

export const BasicValidationCheck = Template.bind({});
BasicValidationCheck.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const submitBtn = canvas.getByRole('button', { name: 'Publish article' });
  const titleInput = canvas.getByPlaceholderText('My new article');

  await userEvent.click(submitBtn);

  canvas.getAllByText(EPublishArticleErrors.TITLE_EMPTY).forEach((el) => {
    expect(el).toBeInTheDocument();
  });

  await userEvent.type(titleInput, 'Sml');
  await userEvent.click(submitBtn);

  expect(
    await canvas.findByText(EPublishArticleErrors.TITLE_LENGTH)
  ).toBeInTheDocument();

  await userEvent.type(titleInput, title);
  await userEvent.click(submitBtn);

  expect(
    await canvas.findByText(EPublishArticleErrors.MARKDOWN_EMPTY)
  ).toBeInTheDocument();
};

export const MarkdownValidationTooShort = Template.bind({});
MarkdownValidationTooShort.args = {
  titleValue: 'Mock article title that is valid',
  markdownContentValue: 'Markdown too short',
};
MarkdownValidationTooShort.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const submitBtn = canvas.getByRole('button', { name: 'Publish article' });

  await userEvent.click(submitBtn);
  expect(
    await canvas.findByText(EPublishArticleErrors.MARKDOWN_TOO_SHORT)
  ).toBeInTheDocument();
};

export const ImageMissing = Template.bind({});
ImageMissing.args = {
  titleValue: title,
  markdownContentValue: content,
};
ImageMissing.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const submitBtn = canvas.getByRole('button', { name: 'Publish article' });

  await userEvent.click(submitBtn);

  canvas.getAllByText(EPublishArticleErrors.IMAGE_EMPTY).forEach((el) => {
    expect(el).toBeInTheDocument();
  });
};

// TODO: Spy on get request
export const SuccessfulSubmit = Template.bind({});
SuccessfulSubmit.args = {
  ...ImageMissing.args,
  imageFileValue: mockOriginalFile,
};
SuccessfulSubmit.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  const submitBtn = canvas.getByRole('button', { name: 'Publish article' });
  await userEvent.click(submitBtn);
};

// export const ContentCheck = Template.bind({});
// ContentCheck.play = async (context) => {
//   const canvas = within(context.canvasElement);
//   const contentEl = canvas.getByPlaceholderText('Supports markdown. Yay!');
//   const submitBtn = canvas.getByRole('button', { name: 'Publish article' });

//   console.log(contentEl);

//   await TitleCheck.play(context);

//   expect(canvas.getByText('* Content cannot be empty!')).toBeInTheDocument();

//   expect(
//     await canvas.findByText(
//       '* Content is too short. Atleast 250 chars are needed'
//     )
//   ).toBeInTheDocument();
// };

// export const EditArticleFormExample = () => (
//   <PublishArticleForm
//     markdownContentValue={lorem}
//     title="Mock article title"
//     imageFileValue={britishCat}
//     onSubmit={createArticleSubmit}
//   />
// );

export default {
  title: 'Organisms/Publish Article Form',
  component: PublishArticleForm,
};
