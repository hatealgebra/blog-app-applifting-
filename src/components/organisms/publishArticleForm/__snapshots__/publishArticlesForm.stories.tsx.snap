// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisms/Forms CreateNewArticleFormExample smoke-test 1`] = `
<form class="sc-gXCJSa kpUsKL">
  <div class="sc-hBpgZr blYtEN">
    <h1>
      Create new article
    </h1>
    <button variant="standard"
            colortheme="primary"
            type="submit"
            class="sc-dkjKgF iJlhGR"
    >
      Publish article
    </button>
  </div>
  <div>
    <div class="sc-jSwlEQ eeEroX">
      <label for="Article title">
        Article title
      </label>
      <input placeholder="My new article"
             id="Article title"
             class="sc-hLclGa iSKFPv"
             value="Article title"
      >
    </div>
    <span class="sc-ipMvLY fOtdNV">
    </span>
  </div>
  <div>
    <div class="sc-ezGUZh fpAwaX">
      <label class="upload-image__label">
        Featured Image
      </label>
      <button variant="standard"
              colortheme="primary"
              type="button"
              class="sc-dkjKgF iJlhGR"
      >
        Upload an Image
      </button>
      <input data-testid="image-uploader"
             type="file"
             accept="image/*"
             style="display: none;"
      >
    </div>
    <span class="sc-ipMvLY fOtdNV">
    </span>
  </div>
  <div>
    <div id="simplemde-editor-1-wrapper">
      <textarea id="simplemde-editor-1"
                placeholder="Supports markdown. Yay!"
                style="display: none;"
      >
      </textarea>
      <div class="EasyMDEContainer"
           role="application"
      >
        <div class="editor-toolbar"
             role="toolbar"
        >
          <button class="bold"
                  type="button"
                  title="Bold (Cmd-B)"
                  aria-label="Bold"
                  tabindex="-1"
          >
            <i class="fa fa-bold">
            </i>
          </button>
          <button class="italic"
                  type="button"
                  title="Italic (Cmd-I)"
                  aria-label="Italic"
                  tabindex="-1"
          >
            <i class="fa fa-italic">
            </i>
          </button>
          <button class="heading"
                  type="button"
                  title="Heading (Cmd-H)"
                  aria-label="Heading"
                  tabindex="-1"
          >
            <i class="fa fa-header fa-heading">
            </i>
          </button>
          <i class="separator">
            |
          </i>
          <button class="quote"
                  type="button"
                  title="Quote (Cmd-')"
                  aria-label="Quote"
                  tabindex="-1"
          >
            <i class="fa fa-quote-left">
            </i>
          </button>
          <button class="unordered-list"
                  type="button"
                  title="Generic List (Cmd-L)"
                  aria-label="Generic List"
                  tabindex="-1"
          >
            <i class="fa fa-list-ul">
            </i>
          </button>
          <button class="ordered-list"
                  type="button"
                  title="Numbered List (Cmd-⌥-L)"
                  aria-label="Numbered List"
                  tabindex="-1"
          >
            <i class="fa fa-list-ol">
            </i>
          </button>
          <i class="separator">
            |
          </i>
          <button class="link"
                  type="button"
                  title="Create Link (Cmd-K)"
                  aria-label="Create Link"
                  tabindex="-1"
          >
            <i class="fa fa-link">
            </i>
          </button>
          <button class="image"
                  type="button"
                  title="Insert Image (Cmd-⌥-I)"
                  aria-label="Insert Image"
                  tabindex="-1"
          >
            <i class="fa fa-image">
            </i>
          </button>
          <i class="separator">
            |
          </i>
          <button class="preview no-disable"
                  type="button"
                  title="Toggle Preview (Cmd-P)"
                  aria-label="Toggle Preview"
                  tabindex="-1"
          >
            <i class="fa fa-eye">
            </i>
          </button>
          <button class="side-by-side no-disable no-mobile"
                  type="button"
                  title="Toggle Side by Side (F9)"
                  aria-label="Toggle Side by Side"
                  tabindex="-1"
          >
            <i class="fa fa-columns">
            </i>
          </button>
          <button class="fullscreen no-disable no-mobile"
                  type="button"
                  title="Toggle Fullscreen (F11)"
                  aria-label="Toggle Fullscreen"
                  tabindex="-1"
          >
            <i class="fa fa-arrows-alt">
            </i>
          </button>
          <i class="separator">
            |
          </i>
          <button class="guide no-disable"
                  type="button"
                  title="Markdown Guide"
                  aria-label="Markdown Guide"
                  tabindex="-1"
          >
            <i class="fa fa-question-circle">
            </i>
          </button>
        </div>
        <div class="CodeMirror cm-s-easymde CodeMirror-wrap"
             translate="no"
             style="clip-path: inset(0px);"
        >
          <div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 15px; left: 15px;">
            <textarea autocorrect="off"
                      autocapitalize="off"
                      spellcheck="true"
                      tabindex="0"
                      style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; min-height: 1em; outline: none;"
            >
            </textarea>
          </div>
          <div class="CodeMirror-vscrollbar"
               tabindex="-1"
               cm-not-content="true"
               style="width: 18px; visibility: hidden;"
          >
            <div style="min-width: 1px; height: 0px;">
            </div>
          </div>
          <div class="CodeMirror-hscrollbar"
               tabindex="-1"
               cm-not-content="true"
               style="height: 18px;"
          >
            <div style="height: 100%; min-height: 1px; width: 0px;">
            </div>
          </div>
          <div class="CodeMirror-scrollbar-filler"
               cm-not-content="true"
          >
          </div>
          <div class="CodeMirror-gutter-filler"
               cm-not-content="true"
          >
          </div>
          <div class="CodeMirror-scroll"
               tabindex="-1"
               style="min-height: 300px;"
          >
            <div class="CodeMirror-sizer"
                 style="margin-left: 0px; margin-bottom: 0px; border-right-width: 50px; min-height: 67px; padding-right: 0px; padding-bottom: 0px;"
            >
              <div style="position: relative; top: 0px;">
                <div class="CodeMirror-lines"
                     role="presentation"
                >
                  <div role="presentation"
                       style="position: relative; outline: none;"
                  >
                    <div class="CodeMirror-measure">
                      <pre class="CodeMirror-line-like">
                        <span>
                          xxxxxxxxxx
                        </span>
                      </pre>
                    </div>
                    <div class="CodeMirror-measure">
                    </div>
                    <div style="position: relative; z-index: 1;">
                    </div>
                    <div class="CodeMirror-cursors">
                      <div class="CodeMirror-cursor"
                           style="left: 4px; top: 0px; height: 19.5px;"
                      >
                        &nbsp;
                      </div>
                    </div>
                    <div class="CodeMirror-code"
                         role="presentation"
                    >
                      <pre class=" CodeMirror-line "
                           role="presentation"
                      >
                        <span role="presentation"
                              style="padding-right: 0.1px;"
                        >
                          Lorem ipsum dolor sit amet consectetur adipisicing elit. Voluptatum optio nesciunt at asperiores cumque quis, enim, ratione repellendus aliquam perferendis nihil atque facere adipisci nulla neque vel beatae eaque eos. lorem onsectetur adipisicing elit. Voluptatum optio nesciunt at asperiores cumque quis, enim, ratione repellendus aliquam perferendis nihil atque facere adipisci nulla neque vel beatae eaqu
                        </span>
                      </pre>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div style="position: absolute; height: 50px; width: 1px; border-bottom: 0px solid transparent; top: 67px;">
            </div>
            <div class="CodeMirror-gutters"
                 style="display: none; height: 117px;"
            >
            </div>
          </div>
        </div>
        <div class="editor-preview-side editor-preview">
        </div>
        <div class="editor-statusbar">
          <span class="autosave">
          </span>
          <span class="lines">
            1
          </span>
          <span class="words">
            55
          </span>
          <span class="cursor">
            1:1
          </span>
        </div>
      </div>
    </div>
    <span class="sc-ipMvLY fOtdNV">
    </span>
  </div>
</form>
`;
